parlai/agents/image_seq2seq/modules.py

```diff
@@ -228,14 +265,80 @@ def forward(
             A (full_enc, full_mask) tuple, which represents the encoded context
             and the mask
         """
+        if self.fusion is FusionType.LATE:
+            return self._forward_late_fusion(src_tokens, image_features)
+        elif self.fusion is FusionType.EARLY:
+            return self._forward_early_fusion(src_tokens, image_features)
+
+        raise RuntimeError(f'Unsupported fusion type: {self.fusion}')
+
+    def _forward_early_fusion(
+        self,
+        src_tokens: Optional[torch.LongTensor],
+        image_features: Optional[Union[List[object], torch.Tensor]],
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """
+        Encode images with context.
+
+        Performs early fusion, whereby image embeddings and token embeddings are computed
+        before passing into the Transformer.
+
+        Essentially overrides normal TransformerEncoder forward.
+        """
+        context_tensor = context_mask = None
+        image_tensor = image_mask = None
+        if src_tokens is not None and image_features is not None:
+            assert src_tokens.size(0) == len(image_features)
+        if src_tokens is not None:
+            context_tensor, context_mask = self.forward_embedding(
+                src_tokens, segments=torch.zeros_like(src_tokens)  # type: ignore
+            )
+        if image_features is not None:
+            valid_img = [v for v in image_features if isinstance(v, torch.Tensor)][0]
+            image_tensor, image_mask = self.embed_images(
+                image_features,
+                segments=torch.ones(  # type: ignore
+                    len(image_features), dtype=torch.long, device=valid_img.device
+                ),
+            )
+
+        # perform early fusion
+        tensor = self._cat([context_tensor, image_tensor])
+        mask: torch.BoolTensor = self._cat([context_mask, image_mask])  # type: ignore
+
+        # Below copied from TransformerEncoder.forward
```

Comment:
Hmm if it's copied should we extract this out into a method of `TransformerEncoder`, perhaps? Otherwise this could quickly diverge from `TransformerEncoder` (since, for instance, I imagine Stephen likely changed this code recently given his layer norm tweak)

Reply:
This is a great point, thank you for bringing it up. 

We've already made steps to factor out the `TransformerEncoder` forward to three "steps" --> embedding input, forwarding through layers, and reducing layer outputs. we *could* further separate into *just* two steps - 1) embedding input and 2) "everything else", but it might obfuscate things a bit more. 

As you mention, we run a risk that something is wrong with the layer norms, however, the specific change you mention does not affect this code directly. Still, I'll add a warning comment here about the layer norms - I would hope that any other fundamental changes to the encoder's forward functions will be implemented in `forward_layers` or `reduce_output` anyway, which is fine in our setup